### PR TITLE
Add `threadid()` and `nthreads()`

### LINKED ISSUE #223: provide Threads.nthreads(), Threads.threadid()
We already have a `Compat.Threads` module, and it would be good to include more functions.   In particular, `nthreads()` and `threadid()` are useful for defining thread-local variables, and can default to `1` on older versions.

### PR DESCRIPTION
**Pull request by ranjanan**
Fixes #223 .

### CODE REVIEW

File: src/Compat.jl [hunk 1]
```diff
@@ -1078,7 +1078,9 @@ if !isdefined(Base, :Threads)
         macro threads(expr)
             return esc(expr)
         end
-        export @threads
+        const threadid() = 1
+        const nthreads() = 1
```
Comment (yuyichao) [inline src/Compat.jl]: No const

File: test/runtests.jl [hunk 2]
```diff
@@ -1142,6 +1142,12 @@ using Compat.Threads
     @test true
 end
 
+# Issue #223
+if VERSION < v"0.5"
```
Comment (yuyichao) [inline test/runtests.jl]: Fwiw this condition is always true even on master.

Comment (ranjanan) [inline test/runtests.jl]: Right, sorry. I'll change that to VERSION > v"0.4"

Comment (ranjanan) [inline test/runtests.jl]: Yeah, I'll change this to `VERSION < v"0.5.0"

**Comment by yuyichao** [comment]
Maybe aa well just relax the test

**Comment by ranjanan** [comment]
@yuyichao You meant not have the test? Because in v0.5, you can start Julia with more than one thread. Then Compat tests would fail.

**Comment by yuyichao** [comment]
No i mean sth like just making sure both exist and 0 < id < n

**Comment by ranjanan** [comment]
Umm, but what would `n` be? Shall I simply say `nthreads() >= 1` and `threadid() == 1` and relax the condition?

**Comment by yuyichao** [comment]
n was short for `nthreads` (and the second `<` was supposed to be `<=`) we can do `1 == threadid() <= nthreads()` too if we want to guarantee that the test will be run on the first thread.

**Comment by yuyichao** [comment]
Also, I think `v"0.5.0" === v"0.5"`

**Comment by ranjanan** [comment]
Okay, I'll simply remove the condition and do `1 == threadid() >= nthreads()`

**Comment by ranjanan** [comment]
@yuyichao I made the same mistake again. Could you please switch off the CI for the previous commits and enable it for the latest commit? Thanks.

**Comment by yuyichao** [comment]
I canceled the extra travis build. I can't do that for AppVeyor ones but I think that one has auto detection.

Anyway, this LGTM now.

File: test/runtests.jl [hunk 3]
```diff
@@ -1142,6 +1142,9 @@ using Compat.Threads
     @test true
 end
 
+# Issue #223
+@test 1 == threadid() <= nthreads()
+
```
Comment (yuyichao) [inline test/runtests.jl]: Need import or qualified name.

Comment (ranjanan) [inline test/runtests.jl]: I think the bug is in a previous line.

File: src/Compat.jl [hunk 4]
```diff
@@ -1078,7 +1078,9 @@ if !isdefined(Base, :Threads)
         macro threads(expr)
             return esc(expr)
         end
-        export @threads
+        threadid() = 1
+        nthreads() = 1
+        export @threads, threadid, threadid
```
Comment (ranjanan) [inline src/Compat.jl]: Yeah, this is the bug. The second `threadid` should be `nthreads`

**Comment by ranjanan** [comment]
It seems `tryparse(Int32, "0", 1)` does not throw an error. It used to throw an `ArgumentError` on v0.4? Is this expected behavior on v0.5?

**Comment by yuyichao** [comment]
Probably https://github.com/JuliaLang/julia/pull/17078 @quinnj

**Comment by ranjanan** [comment]
@yuyichao I see that #17333 has been merged. Could you restart the build for this PR please?

**Comment by yuyichao** [comment]
I can't restart the appveyor here so it's easier if you can rebase this pr.

**Comment by TotalVerb** [comment]
New failure should be fixed by #249.

**Pull request merged**